base/strings/util.jl

```diff
@@ -388,13 +388,83 @@ _replace(io, repl, str, r, pattern) = print(io, repl)
 _replace(io, repl::Function, str, r, pattern) =
     print(io, repl(SubString(str, first(r), last(r))))
 
-replace(str::String, pat_repl::Pair{Char}; count::Integer=typemax(Int)) =
-    replace(str, equalto(first(pat_repl)) => last(pat_repl); count=count)
-replace(str::String, pat_repl::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}};
-        count::Integer=typemax(Int)) =
-    replace(str, occursin(first(pat_repl)) => last(pat_repl), count)
+# replace Char with function that compares with this Char
+pairmap(p::Pair{Char}) = equalto(p.first) => p.second
+
+# replace collection of Char with function that checks occurrence in this collection
+function pairmap(p::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}})
+    occursin(p.first) => p.second
+end
+pairmap(p::Pair) = p
+
+# find the first of a list of patterns.
+# if several match at the same start position prefer the longer one
+# if also the length is equal, prefer the first in list
+# return value of the successful search and the pair
+
+function findnextall(pairs::NTuple{N,Pair}, s::AbstractString, st::Integer=1) where N
+    ps = start(pairs)
```

Comment:
For tuples start is always `1`. You don't need to call `next` either, you can just do `for i in 1:N`.

Reply:
The first I will include the as an optimization.
I want to continue with `while/done/next` because it looks more elegant and performant to me than
```
N = length(pairs)
for i = 1:N
    p = pairs[i]
    ....
    if fr > 0
        ...
        for j = i+1:N
            p = pairs[j]
```

Reply:
More performant: definitely not. More elegant: can always be discussed, but it's not obvious. Note that `N` is already defined as a type parameter.

Reply:
ok, you convinced me. I copied my pattern from somewhere in base or stdlib (don't remember where exactly), and thought performance was the driver.
Thanks for the hint regarding  `N`.
Is it better to introduce a second variable `j` for the inner loop or can I re-use the `i`  ?

Reply:
I'd say it's clearer to use `j`.

Using `next`/`done` is needed to support generic iterators, so that might be the reason why it was used.